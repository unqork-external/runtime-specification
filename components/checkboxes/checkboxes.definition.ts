import { Const, Default, Description, DiscriminatorValue, Optional } from '@tsed/schema'

import { CheckboxesField } from './checkboxes.field'
import { CheckboxesOptions } from './checkboxes.options'
import { CheckboxesStyling } from './checkboxes.styling'
import { CheckboxesTargets } from './checkboxesTargets.enum'
import { ViewTargets } from '../../../decorators'
import { BaseComponentDefinition } from '../../base-component-interface/base.component.definition'
import { Display } from '../../component-composition/display/component.display'
import type { SignalTargets } from '../../signals'

@DiscriminatorValue('checkboxes')
@ViewTargets(CheckboxesTargets)
export class CheckboxesComponentDefinition extends BaseComponentDefinition {
  @Const('checkboxes')
  type: 'checkboxes' = 'checkboxes' as const

  @Optional()
  display: Display = new Display()

  @Optional()
  field: CheckboxesField = new CheckboxesField()

  @Optional()
  options: CheckboxesOptions = new CheckboxesOptions()

  @Optional()
  @Default([])
  @Description('Store values for selected checkboxes')
  value: string[] = []

  @Optional()
  declare styling?: CheckboxesStyling

  @Optional()
  declare signals?: SignalTargets<CheckboxesTargets>
}
