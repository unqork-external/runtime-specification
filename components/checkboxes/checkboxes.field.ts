import { Default, Description, Integer, Optional } from '@tsed/schema'

import { InputField, InputSubLabel } from '../../component-composition/input/component.input'

export class CheckboxesField extends InputField {
  @Optional()
  @Default(1)
  @Integer()
  @Description('Number of columns to split checkboxes')
  numberOfColumns: number = 1

  @Optional()
  subLabel: InputSubLabel = new InputSubLabel()
}
