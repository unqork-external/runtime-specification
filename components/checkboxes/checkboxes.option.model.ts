import { Description, Enum, Required } from '@tsed/schema'

import { CheckboxesOptionType } from './checkboxesOptionType.enum'

export class CheckboxesOptionModel {
  @Required()
  @Description('Option label')
  label: string

  @Required()
  @Enum(CheckboxesOptionType)
  @Description('Option Type')
  type: CheckboxesOptionType

  @Required()
  @Description('Option Value')
  value: string
}
