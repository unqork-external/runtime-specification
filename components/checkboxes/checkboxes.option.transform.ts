import { Default, Description, Example, Optional } from '@tsed/schema'

@Example({ valueLocation: 'value', labelLocation: 'label' })
@Description('Location of value and label for the options.')
export class CheckboxesTransformOptions {
  @Optional()
  @Default('value')
  @Description('Location of value.')
  valueLocation = 'value'

  @Optional()
  @Default('label')
  @Description('Location of label.')
  labelLocation = 'label'
}
