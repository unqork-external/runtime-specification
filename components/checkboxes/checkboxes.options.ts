import { CollectionOf, Default, Description, Optional } from '@tsed/schema'

import { CheckboxesOptionModel } from './checkboxes.option.model'
import { CheckboxesReferenceKeyModel } from './checkboxes.referenceKey.model'

@Description('Options which are taken from the data source')
export class CheckboxesOptions {
  @Optional()
  @Description('Any custom CSS classes to apply to the component.')
  customClass?: string

  @Optional()
  @Description('Data reference key settings for the checkboxes.')
  referenceKey?: CheckboxesReferenceKeyModel

  @Optional()
  @Default(false)
  @Description('Used in renderer - Is None checkbox toggled on')
  isNoneChecked: boolean = false

  @Optional()
  @Default(false)
  @Description('Used in renderer - Is Select All checkbox toggled on')
  isSelectAllChecked: boolean = false

  @Optional()
  @Description('Label property for data reference')
  labelProperty?: string

  @Optional()
  @Default('None of the above')
  @Description('Label for none checkbox')
  noneLabel: string = 'None of the above'

  @Optional()
  @Default('None')
  @Description('Value for none checkbox')
  noneValue: string = 'None'

  @Optional()
  @Default('Select All')
  @Description('Label for select all checkbox')
  selectAllLabel: string = 'Select All'

  @Optional()
  @Default(false)
  @Description('Show none checkbox')
  showNone: boolean = false

  @Optional()
  @Default(false)
  @Description('Show select all checkbox')
  showSelectAll: boolean = false

  @Optional()
  @Description('Value property for data reference source')
  valueProperty?: string

  @Optional()
  @CollectionOf(CheckboxesOptionModel)
  @Default([])
  @Description('Values from the data source')
  values: CheckboxesOptionModel[] = []

  @Optional()
  @CollectionOf(CheckboxesOptionModel)
  @Description('Initial values from the data source')
  initial?: CheckboxesOptionModel[]

  @Optional()
  @Default(false)
  @Description(
    "If true, 'options' exist at the top level of submission data, pointing to the boolean value of their selection status",
  )
  promoteDataToTopLevel: boolean = false
}
