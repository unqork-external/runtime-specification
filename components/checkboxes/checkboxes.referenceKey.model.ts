import { Description, Optional, Required } from '@tsed/schema'

import { CheckboxesTransformOptions } from './checkboxes.option.transform'

export class CheckboxesReferenceKeyModel {
  @Required()
  @Description('The target entity to refer to.')
  target: string

  @Required()
  @Description('The JSONPath representation of what JSON key want to read from.')
  property: string

  @Optional()
  transformOptions: CheckboxesTransformOptions = new CheckboxesTransformOptions()
}
