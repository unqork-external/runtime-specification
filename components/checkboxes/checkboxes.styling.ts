import { Description, Property } from '@tsed/schema'

import { CheckboxesTargets } from './checkboxesTargets.enum'
import { Css } from '../../../decorators/schema/css.decorator'
import { type CssClassOrProps } from '../../styling/cssClassOrProps.type'
import { StylingModel } from '../../styling/styling.model'
import { TargetsModel } from '../../viewTargets/targets.model'

export class CheckboxesStyleTargets implements TargetsModel<CheckboxesTargets> {
  @Css()
  @Description('Adds styles to checkboxes root ComponentWrapper.')
  [CheckboxesTargets.root]: CssClassOrProps;

  @Css()
  @Description('Adds styles to checkboxes read only component.')
  [CheckboxesTargets.readOnly]: CssClassOrProps;

  @Description('Adds styles to Checkboxes label element.')
  @Css()
  [CheckboxesTargets.label]: CssClassOrProps;

  @Description('Adds styles to Checkboxes sub-label element.')
  @Css()
  [CheckboxesTargets.subLabel]: CssClassOrProps;

  @Description('Adds styles to Checkboxes tooltip button element.')
  @Css()
  [CheckboxesTargets.tooltipButton]: CssClassOrProps;

  @Description('Adds styles to Checkboxes tooltip popover element.')
  @Css()
  [CheckboxesTargets.tooltipPopover]: CssClassOrProps;

  @Description('Adds styles to Checkboxes helper text element.')
  @Css()
  [CheckboxesTargets.helperText]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox columns wrapper element.')
  @Css()
  [CheckboxesTargets.checkboxColumnsWrapper]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox column elements.')
  @Css()
  [CheckboxesTargets.checkboxColumns]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox wrapper elements.')
  @Css()
  [CheckboxesTargets.checkboxWrapper]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox label wrapper elements.')
  @Css()
  [CheckboxesTargets.checkboxLabelWrapper]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox label elements.')
  @Css()
  [CheckboxesTargets.checkboxLabel]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox label text elements.')
  @Css()
  [CheckboxesTargets.checkboxLabelText]: CssClassOrProps;

  @Description('Adds styles to Checkboxes checkbox elements.')
  @Css()
  [CheckboxesTargets.checkbox]: CssClassOrProps;

  @Css()
  @Description('Adds styles to Errors component root, which is rendered within the Checkboxes component.')
  [CheckboxesTargets.errorsRoot]: CssClassOrProps;

  @Css()
  @Description('Adds styles to Errors component UL element.')
  [CheckboxesTargets.errorsUnorderedList]: CssClassOrProps;

  @Css()
  @Description('Adds styles to Errors component LI elements.')
  [CheckboxesTargets.errorsUnorderedListItem]: CssClassOrProps;

  @Css()
  @Description('Adds styles to Errors Field error root element, which is rendered within the LI elements.')
  [CheckboxesTargets.errorsFieldError]: CssClassOrProps
}

export class CheckboxesStyling extends StylingModel<CheckboxesStyleTargets> {
  @Property()
  targets: CheckboxesStyleTargets
}
