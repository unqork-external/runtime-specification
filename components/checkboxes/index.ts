export * from './checkboxes.definition'
export * from './checkboxesInputField'
export * from './checkboxesOptions'
