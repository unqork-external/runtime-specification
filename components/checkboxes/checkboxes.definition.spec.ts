import { CheckboxesComponentDefinition } from './checkboxes.definition'
import { generateSchemaAndValidate } from '../../../utilities'

describe('Checkboxes Contract Tests', function () {
  it('should ensure the contract is clear', function () {
    const { schema } = generateSchemaAndValidate(CheckboxesComponentDefinition)

    expect(schema.required).toMatchSnapshot()
    expect(schema.properties).toMatchSnapshot()
    expect(schema.definitions).toMatchSnapshot()
  })
})
