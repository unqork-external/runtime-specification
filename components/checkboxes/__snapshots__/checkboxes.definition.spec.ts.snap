// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkboxes Contract Tests should ensure the contract is clear 1`] = `
[
  "key",
  "type",
]
`;

exports[`Checkboxes Contract Tests should ensure the contract is clear 2`] = `
{
  "display": {
    "$ref": "#/definitions/Display",
  },
  "executable": {
    "default": true,
    "description": "Defines whether or not the object will respond to events or fire operations. When \`true\`, the object will perform and behave as defined in its spec. When \`false\`, the object will no longer be responsive to the system.",
    "type": "boolean",
  },
  "field": {
    "$ref": "#/definitions/CheckboxesField",
  },
  "inputs": {
    "items": {
      "$ref": "#/definitions/InputRef",
    },
    "type": "array",
  },
  "key": {
    "description": "'key' is the user-defined, unique identifier for an object within a module. It doesn't have to be unique across different modules. Objects within the runtime use keys to target other objects.",
    "examples": [
      "I am the key",
      "bestKeyEver",
    ],
    "minLength": 1,
    "type": "string",
  },
  "options": {
    "$ref": "#/definitions/CheckboxesOptions",
  },
  "package": {
    "$ref": "#/definitions/PackageModel",
  },
  "signals": {
    "description": "An object which contains all of the signals for the component. The signals object can have any number of target keys as defined in the component definition which each contain an array of Signals.",
    "examples": [
      {
        "args": {
          "inputs": {
            "propertyPath": "value",
            "required": true,
            "targetKey": "firstName",
          },
        },
        "id": "firstNameWatcher",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "firstNameWatcher operations fired: value watcher triggered",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onWatch",
      },
      {
        "id": "textfieldOnBlur",
        "operations": [
          {
            "options": {
              "logType": "warn",
              "thingToPrint": "textfield received blur event from DOM",
            },
            "type": "PRINTER",
          },
        ],
        "type": "onBlur",
      },
      "<Example> Prompt: "I want a textfield that will trigger the root execute signal on myValidationComponent when the textfield input receives a blur event." Response: [{"key":"yourGeneratedTextfield","type":"textfield","signals":{"input":[{"id":"textfieldOnBlur","operations":[{"type":"EXECUTE","options":{"targetKey":"myValidationComponent","signalTarget":"root","shouldExecute":true}}],"type":"onBlur"}]}}] </Example>",
    ],
    "type": "object",
  },
  "styling": {
    "$ref": "#/definitions/CheckboxesStyling",
  },
  "type": {
    "const": "checkboxes",
    "description": "Name of the type of component.",
    "examples": [
      "checkboxes",
    ],
    "minLength": 1,
    "type": "string",
  },
  "validation": {
    "$ref": "#/definitions/Validation",
  },
  "value": {
    "description": "Store values for selected checkboxes",
    "items": {
      "default": [],
    },
    "type": "array",
  },
}
`;

exports[`Checkboxes Contract Tests should ensure the contract is clear 3`] = `
{
  "CheckboxesField": {
    "description": "Details to be applied to the input element within a component.",
    "properties": {
      "customClass": {
        "description": "Adds a custom CSS class to the component. This is useful for targeting this component to apply custom styles.",
        "type": "string",
      },
      "helperText": {
        "$ref": "#/definitions/InputHelperText",
      },
      "label": {
        "$ref": "#/definitions/Label",
      },
      "numberOfColumns": {
        "default": 1,
        "description": "Number of columns to split checkboxes",
        "multipleOf": 1,
        "type": "integer",
      },
      "placeholder": {
        "description": "An example or short explanation of the requested input. This appears inside of the input by default and disappears when the user enters a value.",
        "examples": [
          "First name - Expects the user to type their first name",
          "Enter your Social Security Number - Expects user's SSN",
          "dd/MM/YYYY - To show an expected date format (if you need a more robust component, try Date Input",
        ],
        "type": "string",
      },
      "prefix": {
        "description": "A symbol or text to display before the input field. Note this value will not be part of the submitted data",
        "examples": [
          "$",
          "@",
        ],
        "type": "string",
      },
      "subLabel": {
        "$ref": "#/definitions/InputSubLabel",
      },
      "suffix": {
        "description": "A symbol or text to display after the input field. Note this value will not be part of the submitted data",
        "examples": [
          ".com",
        ],
        "type": "string",
      },
      "tooltipDescription": {
        "description": "Tooltips are floating texts that appear when the cursor is positioned over an icon. They can span across multiple lines and provide hints about the requested data. Setting this property will make a question mark icon to be displayed on the right side of the component's label, and once the user hovers over it, the tooltip text will appear. Moving the cursor outside the icon's area hides the tooltip. Hiding the component's label automatically hides the tooltip.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "CheckboxesOptionModel": {
    "properties": {
      "label": {
        "description": "Option label",
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "description": "Option Type",
        "enum": [
          "value",
          "selectAll",
          "none",
        ],
        "minLength": 1,
        "type": "string",
      },
      "value": {
        "description": "Option Value",
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "label",
      "type",
      "value",
    ],
    "type": "object",
  },
  "CheckboxesOptions": {
    "description": "Options which are taken from the data source",
    "properties": {
      "customClass": {
        "description": "Any custom CSS classes to apply to the component.",
        "type": "string",
      },
      "initial": {
        "description": "Initial values from the data source",
        "items": {
          "$ref": "#/definitions/CheckboxesOptionModel",
        },
        "type": "array",
      },
      "isNoneChecked": {
        "default": false,
        "description": "Used in renderer - Is None checkbox toggled on",
        "type": "boolean",
      },
      "isSelectAllChecked": {
        "default": false,
        "description": "Used in renderer - Is Select All checkbox toggled on",
        "type": "boolean",
      },
      "labelProperty": {
        "description": "Label property for data reference",
        "type": "string",
      },
      "noneLabel": {
        "default": "None of the above",
        "description": "Label for none checkbox",
        "type": "string",
      },
      "noneValue": {
        "default": "None",
        "description": "Value for none checkbox",
        "type": "string",
      },
      "promoteDataToTopLevel": {
        "default": false,
        "description": "If true, 'options' exist at the top level of submission data, pointing to the boolean value of their selection status",
        "type": "boolean",
      },
      "referenceKey": {
        "$ref": "#/definitions/CheckboxesReferenceKeyModel",
      },
      "selectAllLabel": {
        "default": "Select All",
        "description": "Label for select all checkbox",
        "type": "string",
      },
      "showNone": {
        "default": false,
        "description": "Show none checkbox",
        "type": "boolean",
      },
      "showSelectAll": {
        "default": false,
        "description": "Show select all checkbox",
        "type": "boolean",
      },
      "valueProperty": {
        "description": "Value property for data reference source",
        "type": "string",
      },
      "values": {
        "description": "Values from the data source",
        "items": {
          "$ref": "#/definitions/CheckboxesOptionModel",
        },
        "type": "array",
      },
    },
    "type": "object",
  },
  "CheckboxesReferenceKeyModel": {
    "properties": {
      "property": {
        "description": "The JSONPath representation of what JSON key want to read from.",
        "minLength": 1,
        "type": "string",
      },
      "target": {
        "description": "The target entity to refer to.",
        "minLength": 1,
        "type": "string",
      },
      "transformOptions": {
        "$ref": "#/definitions/CheckboxesTransformOptions",
      },
    },
    "required": [
      "target",
      "property",
    ],
    "type": "object",
  },
  "CheckboxesStyleTargets": {
    "properties": {
      "checkbox": {
        "description": "Adds styles to Checkboxes checkbox elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxColumns": {
        "description": "Adds styles to Checkboxes checkbox column elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxColumnsWrapper": {
        "description": "Adds styles to Checkboxes checkbox columns wrapper element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxLabel": {
        "description": "Adds styles to Checkboxes checkbox label elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxLabelText": {
        "description": "Adds styles to Checkboxes checkbox label text elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxLabelWrapper": {
        "description": "Adds styles to Checkboxes checkbox label wrapper elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "checkboxWrapper": {
        "description": "Adds styles to Checkboxes checkbox wrapper elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "errorsFieldError": {
        "description": "Adds styles to Errors Field error root element, which is rendered within the LI elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "errorsRoot": {
        "description": "Adds styles to Errors component root, which is rendered within the Checkboxes component.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "errorsUnorderedList": {
        "description": "Adds styles to Errors component UL element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "errorsUnorderedListItem": {
        "description": "Adds styles to Errors component LI elements.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "helperText": {
        "description": "Adds styles to Checkboxes helper text element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "label": {
        "description": "Adds styles to Checkboxes label element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "readOnly": {
        "description": "Adds styles to checkboxes read only component.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "root": {
        "description": "Adds styles to checkboxes root ComponentWrapper.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "subLabel": {
        "description": "Adds styles to Checkboxes sub-label element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "tooltipButton": {
        "description": "Adds styles to Checkboxes tooltip button element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
      "tooltipPopover": {
        "description": "Adds styles to Checkboxes tooltip popover element.",
        "oneOf": [
          {
            "type": "string",
          },
          {
            "type": "object",
          },
          {
            "items": {
              "oneOf": [
                {
                  "type": "string",
                },
                {
                  "type": "object",
                },
              ],
            },
            "type": "array",
          },
        ],
      },
    },
    "type": "object",
  },
  "CheckboxesStyling": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "$ref": "#/definitions/CheckboxesStyleTargets",
        "type": "object",
      },
    },
    "type": "object",
  },
  "CheckboxesTransformOptions": {
    "description": "Location of value and label for the options.",
    "examples": [
      {
        "labelLocation": "label",
        "valueLocation": "value",
      },
    ],
    "properties": {
      "labelLocation": {
        "default": "label",
        "description": "Location of label.",
        "type": "object",
      },
      "valueLocation": {
        "default": "value",
        "description": "Location of value.",
        "type": "object",
      },
    },
    "type": "object",
  },
  "Display": {
    "description": "Basic UI settings related to how a component may be displayed and interacted with. 'Hidden' corresponds to visibility. 'Interactive' corresponds to the end-user's ability to interact with input components. 'Read Only View' corresponds to the component being rendered with an input element or as text.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "Setting it to \`true\` hides the component from the view, while \`false\` shows it. Regardless of this setting, the component is always visible in the Module Editor. A hidden component can still receive information from other components.",
        "type": "boolean",
      },
      "interactive": {
        "default": true,
        "description": "When \`false\`, it is not possible to interact with or edit the input field. Its appearance is modified (grayed-out by default) to show its state, the HTML 'disabled' attribute is added to the field, and when hovering over it, the cursor also changes to alert of the disabled status. However, the component can still receive information from other components when it's not interactive.",
        "type": "boolean",
      },
      "readOnlyView": {
        "default": false,
        "description": "Defines how the component will be displayed. When \`false\`, the component is displayed as an input field. When \`true\`, the component will not accept any input from the user, and its value will be rendered inline as text and not as an input component. If no value is present, the word 'None' is rendered. If a prefix or suffix is set for the component, they also appear as inline text.",
        "type": "boolean",
      },
    },
    "type": "object",
  },
  "InputHelperText": {
    "description": "Details pertaining to the helper text element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "When \`true\`, it hides the helper text of the component.",
        "type": "boolean",
      },
      "text": {
        "description": "Configurable text that is displayed underneath the input field. Unlike tooltips, a helper text stays visible under the input field and displays on a single line.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputLabel": {
    "description": "Details pertaining to the label element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "When \`true\`, it hides the label of the component. Note the label will still be displayed on the Module Editor.",
        "type": "boolean",
      },
      "text": {
        "description": "Configurable text that is displayed above the input field.",
        "type": "string",
      },
    },
    "type": "object",
  },
  "InputRef": {
    "description": "Objects may define other targets that they refer to. This becomes a link for operation execution within the Runtime.",
    "properties": {
      "alias": {
        "type": "string",
      },
      "propertyPath": {
        "type": "string",
      },
      "required": {
        "type": "boolean",
      },
      "silent": {
        "type": "boolean",
      },
      "targetKey": {
        "minLength": 1,
        "type": "string",
      },
      "type": {
        "type": "string",
      },
    },
    "required": [
      "targetKey",
    ],
    "type": "object",
  },
  "InputSubLabel": {
    "description": "Details pertaining to the sub-label element on a component, including the text to be displayed and visibility.",
    "properties": {
      "hidden": {
        "default": false,
        "description": "Hide sub-label text of the component",
        "type": "boolean",
      },
      "text": {
        "description": "Sub-label text for the component",
        "type": "string",
      },
    },
    "type": "object",
  },
  "Label": {
    "properties": {
      "text": {
        "description": "Text displayed as the label for the field",
        "type": "string",
      },
    },
    "type": "object",
  },
  "PackageModel": {
    "description": "The PackageModel describes how the Vega Runtime will resolve dependencies",
    "properties": {
      "type": {
        "description": "Type of package",
        "type": "string",
      },
      "url": {
        "description": "URL of package",
        "type": "string",
      },
      "version": {
        "description": "Version of package",
        "type": "string",
      },
    },
    "type": "object",
  },
  "StylingModel": {
    "description": "Holds information for targeted styling of the component.",
    "properties": {
      "targets": {
        "type": "object",
      },
    },
    "type": "object",
  },
  "Validation": {
    "description": "Holds the validation rules and the most current validation results for the component",
    "properties": {
      "isDirty": {
        "description": "Whether or not field has had user interaction. It will be set to true on user interaction.",
        "type": "boolean",
      },
      "result": {
        "$ref": "#/definitions/ValidationResult",
      },
      "rules": {
        "description": "The set of rules to restrict or define the accepted data for the current component. These are the rules to run for validation of the data input.",
        "examples": [
          "You can use it to define the max length of characters in an input field with a custom error message. rules: { required: { type: ValidationRuleTypes.MAX_LENGTH, validation: { maxLength: 5, }, errorMessage: 'Max length is 5', }, },",
        ],
        "type": "object",
      },
    },
    "type": "object",
  },
  "ValidationError": {
    "properties": {
      "message": {
        "description": "The error message",
        "type": "string",
      },
      "type": {
        "description": "Type of the error",
        "enum": [
          "required",
          "maxLength",
          "minLength",
          "arrayMinLength",
          "arrayMaxLength",
          "pattern",
          "min",
          "max",
          "mask",
          "number",
          "dateinput",
          "isBeforeDate",
          "isAfterDate",
          "customError",
          "isEmail",
        ],
        "minLength": 1,
        "type": "string",
      },
    },
    "required": [
      "type",
    ],
    "type": "object",
  },
  "ValidationResult": {
    "properties": {
      "errors": {
        "items": {
          "$ref": "#/definitions/ValidationError",
        },
        "type": "array",
      },
      "isValid": {
        "type": [
          "null",
          "boolean",
        ],
      },
    },
    "type": "object",
  },
}
`;
